File: server/agent/pkce.js
'use strict';

const crypto = require('crypto');

/**
 * Shared OAuth 2.0 PKCE (S256) primitives used by every OAuth provider in this
 * app (Codex, Claude). Kept provider-agnostic and pure so both flows follow the
 * same oauth-oidc checklist: PKCE S256 only, cryptographically random state.
 */

/** base64url with no padding (RFC 7636 §A). */
function base64url(buf) {
  return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/** PKCE code verifier: 43 chars of base64url (32 random bytes). */
function generateVerifier() {
  return base64url(crypto.randomBytes(32));
}

/** PKCE S256 challenge for a verifier. */
function challengeFromVerifier(verifier) {
  return base64url(crypto.createHash('sha256').update(verifier, 'ascii').digest());
}

/** Cryptographically random, unguessable CSRF state. */
function generateState() {
  return base64url(crypto.randomBytes(32));
}

module.exports = { base64url, generateVerifier, challengeFromVerifier, generateState };

File: ai_fleet/agent/pkce.py
"""Shared OAuth 2.0 PKCE (S256) primitives (port of agent/pkce.js).

Provider-agnostic and pure so both OAuth flows (Codex, Claude) follow the same
checklist: PKCE S256 only, cryptographically random state.
"""

from __future__ import annotations

import base64
import hashlib
import secrets


def base64url(buf: bytes) -> str:
    """base64url with no padding (RFC 7636 §A)."""
    return base64.urlsafe_b64encode(buf).decode("ascii").rstrip("=")


def generate_verifier() -> str:
    """PKCE code verifier: 43 chars of base64url (32 random bytes)."""
    return base64url(secrets.token_bytes(32))


def challenge_from_verifier(verifier: str) -> str:
    """PKCE S256 challenge for a verifier."""
    return base64url(hashlib.sha256(verifier.encode("ascii")).digest())


def generate_state() -> str:
    """Cryptographically random, unguessable CSRF state."""
    return base64url(secrets.token_bytes(32))
